Maple2.Server.Game/Manager/MasteryManager.cs

```diff
@@ -88,6 +94,81 @@ public class MasteryManager {
             return 1;
         }
 
-        return (short) Math.Max(1, masteryRewardEntries.FirstOrDefault(mastery => session.Mastery[MasteryType.Fishing] >= mastery.Value.Value).Key);
+        return (short) Math.Max(1, masteryRewardEntries.OrderByDescending(mastery => mastery.Key).FirstOrDefault(mastery => session.Mastery[type] >= mastery.Value.Value).Key);
+    }
+
+    public void Gather(FieldInteract fieldInteract) {
+        if (!session.TableMetadata.MasteryRecipeTable.Entries.TryGetValue(fieldInteract.Value.Item.RecipeId, out MasteryRecipeTable.Entry? recipeMetadata)) {
+            session.Send(MasteryPacket.Error(MasteryError.s_mastery_error_unknown));
+            return;
+        }
+
+        if (recipeMetadata.RequiredMastery > this[recipeMetadata.Type]) {
+            session.Send(MasteryPacket.Error(MasteryError.s_mastery_error_lack_mastery));
+            return;
+        }
+
+        if (!gatheringCounts.TryGetValue(recipeMetadata.Id, out int currentCount)) {
+            gatheringCounts[recipeMetadata.Id] = currentCount;
+        }
+
+        // TODO: Implement the bool on the last parameter of the formula. Is in someone else's home ?
+        float successRate = lua.CalcGatheringObjectSuccessRate(currentCount, recipeMetadata.HighRateLimitCount, recipeMetadata.NormalRateLimitCount);
+
+        int gatheringAmount = 0;
+        switch (recipeMetadata.Type) {
+            case MasteryType.Farming:
```

Comment:
Can we combine these case statements like Gathering+Mining? Also what's even the difference between harvest/farming?

Reply:
There's almost no difference. It's very weird but each have different condition types.